File: Engimon.cpp
#include "Engimon.h"
//4 sekawan
int Engimon::totalEngimon = 0;
Engimon::Engimon() : id(totalEngimon+1){
    this->setName("None");
    this->setParent(nullptr,nullptr);
    this->setSpecies(Species());
    this->addSkill(Skill());
    this->addSkill(species.getUniqueSkill());
    this->addElement(Element());
    this->setLevel(0);
    this->setExp(0);
    this->setCumulativeExp(0);
    this->setExpLimit(0);
}

Engimon::Engimon(string name, Engimon* parent1, Engimon* parent2, Species species, vector<Skill> skills, vector<Element> elements, int level, int explimit) : id(totalEngimon+1){
    this->setName(name);
    this->setParent(parent1,parent2);
    this->setSpecies(species);
    this->addSkill(skills);
    this->addSkill(species.getUniqueSkill());
    this->addElement(elements);
    this->setLevel(level);
    this->setExp(0);
    this->setCumulativeExp(this->getLevel()*100);
    this->setExpLimit(explimit);
}

Engimon::Engimon(const Engimon& e) : id(e.id){
    this->setName(e.name);
    this->setParent(e.parents[0],e.parents[1]);
    this->setSpecies(e.species);
    this->addSkill(e.skills);
    this->addElement(e.elements);
    this->setLevel(level);
    this->setExp(0);
    this->setCumulativeExp(this->getLevel()*100);
    this->setExpLimit(e.explimit);
}

Engimon::~Engimon(){}

Engimon& Engimon::operator=(const Engimon& e){
    this->setName(e.name);
    this->setParent(e.parents[0],e.parents[1]);
    this->setSpecies(e.species);
    this->addSkill(e.skills);
    this->addElement(e.elements);
    this->setLevel(level);
    this->setExp(0);
    this->setCumulativeExp(this->getLevel()*100);
    this->setExpLimit(e.explimit);
    return *this;
}

//setter
void Engimon::setName(string name){
    this->name = name;
}
void Engimon::setParent(Engimon* e1, Engimon* e2){
    this->parents.push_back(e1);
    this->parents.push_back(e2);
}
void Engimon::setSpecies(Species s){
    this->species = s;
}
void Engimon::addSkill(Skill s){
    if (this->skills.size() < 4 ){
        if (this->hasSkill(s)==false){
            this->skills.push_back(s);
            sort(this->skills.begin(), this->skills.end());
            reverse(this->skills.begin(), this->skills.end());
        }
        else{
            //exception : sudah punya skill tsb
            //kayanya ngga usah diimplemen, biar simple
        }
    }
    else{
        //exception : skill slot full
        this->replaceSkill(s);
    }
}
void Engimon::addSkill(vector<Skill> vs){
    for (int i=0; i<vs.size();i++){
        this->addSkill(vs[i]);
    }
}
void Engimon::addElement(Element e){
    if (this->isElement(e) == false){
        this->elements.push_back(e);
    }
}
void Engimon::addElement(vector<Element> ve){
    for (int i=0; i<ve.size();i++){
        this->addElement(ve[i]);
    }
}
void Engimon::levelUp(){
    this->exp -= 100;
    this->level++;
}
void Engimon::setLevel(int lvl){
    this->level = lvl;
}
void Engimon::addExp(int exp){
    this->exp += exp;
    this->cumulativeExp +=exp;
    while (this->lvlUpEligibility()){
        this->levelUp();
    }
    if (this->isDead()){
        //idk, maybe some exception or something?
    }
}
void Engimon::setExp(int exp){
    this->exp = exp;
}
void Engimon::setCumulativeExp(int cumxp){
    this->cumulativeExp = cumxp;
}
void Engimon::setExpLimit(int explimit){
    this->explimit = explimit;
}

//getter
int Engimon::getId(){
    return this->id;
}
string Engimon::getName(){
    return this->name;
}
vector<Engimon*> Engimon::getParents(){
    return this->parents;
}
Species Engimon::getSpecies(){
    return this->species;
}
vector<Skill> Engimon::getSkills(){
    return this->skills;
}
vector<Element> Engimon::getElements(){
    return this->elements;
}
int Engimon::getLevel(){
    return this->level;
}
int Engimon::getExp(){
    return this->exp;
}
int Engimon::getCumulativeExp(){
    return this->cumulativeExp;
}
int Engimon::getCumExpLimit(){
    return this->explimit;
}

Skill Engimon::getHighestMasteryLevel(){
    Skill maxSkill = this->skills[0];
    int max = this->skills[0].getMasteryLevel();
    for (int i = 1; i < this->skills.size() ; i++){
        if (max < this->skills[i].getMasteryLevel()){
            max = this->skills[i].getMasteryLevel();
            maxSkill = this->skills[i];
        }
    }
    return maxSkill;
}

//get status
bool Engimon::lvlUpEligibility(){
    return this->exp>=100;
}
bool Engimon::isDead(){
    return (this->getCumulativeExp() >= this->getCumExpLimit());
}
bool Engimon::isElement(Element e){
    vector<Element> el = this->getElements();
    for (int i=0 ; i<el.size() ; i++){
        if (el[i].getElementName() == e.getElementName()){
            return true;
        }
    }
    return false;
}

bool Engimon::hasSkill(Skill s){
    for (int i = 0 ; i < this->skills.size() ; i++){
        if (this->skills[i].getSkillName() == s.getSkillName()){
            return true;
        }
    }
    return false;
}

//delete
void Engimon::replaceSkill(Skill s) {
    string input;
    int idx;
    cout << "Skill Engimon penuh, Apakah player ingin mengganti skill yang ada? (Y/N) : ";
    do {
        cin >> input;
    } while (input != "Y" && input != "N");
    if (input == "Y") {
        this->printSkill();
        cout << "Pilih skill yang akan diganti (By index) : ";
        do {
            cin >> idx;
        } while (idx < 1 || idx > this->skills.size());
        this->skills.erase(this->skills.begin() + (idx-1));
        this->addSkill(s);
        cout << "Skill telah diganti" << endl;
    }
}

//output func
void Engimon::printDetail(){
    cout<<"Nama Engimon"<<"\t: "<<this->getName()<<endl;
    if (this->getParents()[0] != nullptr && this->getParents()[1]!=nullptr){
        for (int i=0 ; i<2 ; i++){
            cout<<"Parent "<<i+1<<"\t: "<<this->getParents()[i]->getName()<<" ("<<this->getParents()[i]->getSpecies().getSpeciesName()<<")"<<endl;
        }
    }
    cout<<"Spesies"<<"\t\t: "<<this->getSpecies().getSpeciesName()<<endl;
    cout<<"Skill\t\t: ";
    for (int i=0;i<this->getSkills().size();i++){
        if (i<this->getSkills().size()-1){
            cout<<this->getSkills()[i].getSkillName()<<", ";
        }
        else{
            cout<<this->getSkills()[i].getSkillName()<<endl;
        }
        
    }
    cout<<"Element\t\t: ";
    for (int i=0;i<this->getElements().size();i++){
        if (i<this->getElements().size()-1){
            cout<<this->getElements()[i].getElementName()<<", ";
        }
        else{
            cout<<this->getElements()[i].getElementName()<<endl;
        }
        
    }
    cout<<"Level"<<"\t\t: "<<this->getLevel()<<endl;
    cout<<"Experience"<<"\t: "<<this->getExp()<<endl;
    cout<<"Cumulative Exp."<<"\t: "<<this->getCumulativeExp()<<endl;

}

void Engimon::printSkill() {
    cout << "Skill\t\t: " << endl;
    for (int i=0; i<this->getSkills().size(); i++){
        cout << (i+1) << ". " << this->getSkills()[i].getSkillName() << endl;
    }
}

File: willywangky/src/willywank/mainobjects/Engimon.java
package willywank.mainobjects;

import java.util.ArrayList;
import java.util.List;

public class Engimon implements Comparable<Engimon> {
    int id;
    String name;
    List<Engimon> parents = new ArrayList<>();
    Species species;
    List<Skill> skills = new ArrayList<>();
    List<Element> elements = new ArrayList<>();
    int level;
    int exp;
    int cumulativeExp;
    int explimit;
    int health;

    static int totalEngimon = 0;

    public Engimon(){
        setName("None");
        setParent(null,null);
        setSpecies(new Species());
        addElement(new Element());
        setLevel(0);
        setExp(0);
        setCumulativeExp(getLevel()*100 - 100);
        setExpLimit(0);
        this.health = 3;
        totalEngimon++;
    }

    public Engimon(String name, Engimon p1, Engimon p2, Species species, List<Skill> skills, List<Element> elements, int level, int explimit){
        this.id = totalEngimon+1;
        setName(name);
        setParent(p1,p2);
        setSpecies(species);
        addSkill(skills);
        try {
            addSkill(species.getUniqueSkill());
        } catch (SkillNotExist e) {
            // TODO Auto-generated catch block
            e.showErrors();
        } catch (SkillOverload e) {
            // TODO Auto-generated catch block
            e.showErrors();
        }
        addElement(elements);
        setLevel(level);
        setExp(0);
        setCumulativeExp(getLevel()*100 - 100);
        setExpLimit(explimit);
        this.health = 3;
        totalEngimon++;
    }

    public Engimon(Engimon e){
        this.id = e.id;
        setName(e.getName());
        setParent(e.getParents().get(0),e.getParents().get(1));
        setSpecies(e.getSpecies());
        addSkill(e.getSkills());
        try {
            addSkill(e.getSpecies().getUniqueSkill());
        } catch (SkillNotExist exception) {
            // TODO Auto-generated catch block
            exception.showErrors();
        } catch (SkillOverload exception) {
            // TODO Auto-generated catch block
            exception.showErrors();
        }
        addElement(e.getElements());
        setLevel(e.getLevel());
        setExp(0);
        setCumulativeExp(this.getLevel()*100 - 100);
        setExpLimit(e.getCumExpLimit());
        this.health = 3;
        totalEngimon++;
    }

    @Override
    public boolean equals(Object o) {

        // If the object is compared with itself then return true
        if (o == this) {
            return true;
        }

        /* Check if o is an instance of Complex or not
          "null instanceof [type]" also returns false */
        if (!(o instanceof Engimon)) {
            return false;
        }

        // typecast o to Complex so that we can compare data members
        Engimon e = (Engimon) o;

        // Compare the data members and return accordingly
        return (id == e.id);
    }

    @Override
    public int compareTo(Engimon e) {

        if (this.elements.get(0).elementToNumber() == e.getElements().get(0).elementToNumber()) {
            return 0;
        }
        if (this.elements.get(0).elementToNumber() < e.getElements().get(0).elementToNumber()) {
            return -1;
        }
        if (this.elements.get(0).elementToNumber() > e.getElements().get(0).elementToNumber()) {
            return 1;
        }
        return 0;
    }

    public void setName(String name){
        this.name = name;
    }
    public void setParent(Engimon p1, Engimon p2){
        this.parents.add(p1);
        this.parents.add(p2);
    }
    public void setSpecies(Species species){
        this.species = species;
    }
    public void addSkill(Skill skill) throws SkillNotExist, SkillOverload{
        if (this.skills.size() < 4){
            if (!this.hasSkill(skill)){
                this.skills.add(skill);
            }
            else{
                throw new SkillNotExist();
            }
        }
        else{
            throw new SkillOverload();
        }

    }

    public void addSkill(List<Skill> arr){
        for (Skill skill : arr) {
            try {
                this.addSkill(skill);
            } catch (SkillNotExist e) {
                // TODO Auto-generated catch block
                e.showErrors();
            } catch (SkillOverload e) {
                // TODO Auto-generated catch block
                e.showErrors();
            }
        }
    }
    public void addElement(Element E){
        if (!isElement(E))
            this.elements.add(E);
    }
    public void addElement(List<Element> E){
        for (Element element : E) {
            this.addElement(element);
        }
    }
    public void levelUp(){
        this.exp -= 100;
        this.level++;
        System.out.println("\nEngimon Level Up!");
        System.out.println(this.name + ", current level" + this.level);
    }
    public void setLevel(int level){
        this.level = level;
        setCumulativeExp(this.level*100 - 100);
    }
    public void addExp(int exp){
        this.exp += exp;
        this.cumulativeExp +=exp;
        while (lvlUpEligibility()){
            levelUp();
        }
//        if (isDead()){
//            //idk, maybe some exception or something?
//        }
    } //ini berpengaruh ke exp dan cumulative exp
    public void setExp(int exp){
        this.exp = exp;
    }
    public void setCumulativeExp(int cumulative){
        this.cumulativeExp = cumulative;
    } //ini kurang perlu menurutku tapi jaga-jaga aja
    public void setExpLimit(int lim){
        this.explimit = lim;
    }
    public int getId(){
        return this.id;
    }
    public String getName(){
        return this.name;
    }
    public List<Engimon> getParents(){
        return this.parents;
    }
    public Species getSpecies(){
        return this.species;
    }
    public List<Skill> getSkills(){
        return this.skills;
    }
    public List<Element> getElements(){
        return this.elements;
    }
    public int getLevel(){
        return this.level;
    }
    public int getExp(){
        return this.exp;
    }
    public int getCumulativeExp(){
        return this.cumulativeExp;
    }
    public int getCumExpLimit(){
        return this.explimit;
    }
    public int getHealth(){return this.health;}
    Skill getHighestMasteryLevel(){
        Skill maxSkill = this.skills.get(0);
        int max = maxSkill.getMasteryLevel();
        for (Skill skill : this.skills) {
            if (skill.getMasteryLevel() > max){
                max = skill.getMasteryLevel();
                maxSkill = skill;
            }
        }
        return maxSkill;
    }
    public Boolean lvlUpEligibility(){
        return this.exp >= 100;
    } //refer to spek 1.c
    public Boolean isDead(){
        return getCumulativeExp() >= getCumExpLimit() || this.health<=0;
    } //refer to spek 1.d
    public Boolean isElement(Element E){
        return this.elements.contains(E);
    } //apakah engimon berelemen x
    public Boolean hasSkill(Skill S){
        return this.skills.contains(S);
    } //apakah engimon punya skill x
    public void decreaseHealth(){
        this.health--;
    }
    //delete
    void replaceSkill(Skill S){

    } //replace skill by index

    // Update: Passive exp nambah buat engimon liar
    void passiveExp(){
        addExp(100);
    }

    //output function
    void printDetail(){

    }
    void printSkill(){

    }
    @Override
    public String toString() {
        return String.format(this.name+" / "+this.getElements().get(0).getElementName()+" / Lv."+this.getLevel());
    }
}
